### DISCUSSION TITLE
Graphql query for singular data model collection

### DISCUSSION THREAD

**Discussion by yudyananda** [directus/directus#4084] [Archive]
I am used to using wordpress and gatsbyjs with graphql, but currently I am very interested in using Directus for my next website project. The problem is I'm used to plural and singular options like [wpgraphql](https://www.wpgraphql.com/docs/posts-and-pages/). For example let say if I have a data model collection of **posts** then my expectation there will be a **post** as singular version query from **posts**

Is this unfinished feature or something I need to adapt to?

**Comment by rijkvanzanten** [comment]
> Is this unfinished feature or something I need to adapt to?

I'm not sure! There is an open feature request to allow for both singular and plural collection names: https://github.com/directus/directus/discussions/3069, but that hasn't been approved yet. I'm also not sure if that would be used in GraphQL to retrieve singular data. As of right now, it's something you'll have to adapt to 👍🏻

**Reply by rijkvanzanten** [reply]
It would be really helpful if you could outline how you think it should work exactly, and why you think it's useful 🙂 That way, we can keep the discussion going and potentially implement and ship it 😄

**Reply by yudyananda** [reply]
@rijkvanzanten thanks for response. I'm not sure how to get the single piece data from the collection (data model) if there is no singular type option. The only way that i know is `filter` which we'll always get a list even when all we need is a singular one. This will quite frustrating. 

```javascript

// without singular option  
query ($id: ID!) {
     items {
        posts(filter: {id: {_eq: $id}}) {
           id
           title
    }
  }
}

// with singular option

query ($id: ID! {
     items {
        singlePost: post(id: $id) {
           id
           title
    }
  }
}
```
In my opinion this will a lot helpful. I really hope this feature will be considered and approved by you, @benhaynes  or other team members.

**Reply by philleepflorence** [reply]
@rijkvanzanten - Is this related to the **single** parameter?
So in REST, you get a single object not an array of a single object, and then in graphQL, a singular of the plural.

I know it was in D8 for REST, but not sure yet about D9.
[Directus 8 - Single Documentation](https://v8.docs.directus.io/api/query/single.html)

**Reply by rijkvanzanten** [reply]
The `single` is no longer a thing, as changing the output structure based on an attribute is not a thing in GraphQL.

I get that having an object instead of an array of 1 is useful @yudyananda, but I don't agree it's _that_ big of a frustration to just use `result[0]` instead of `result` 🤷🏻

**Reply by benhaynes** [reply]
Not sure if this is relevant, but we did have the idea of setting singular and plural terms to each collection. Though I agree that it seems pretty trivial to return an array and just fetch the first/only item. https://github.com/directus/directus/issues/3227

**Reply by philleepflorence** [reply]
> The **single** is no longer a thing...

It may be helpful to add a reference or link to a resource for instances like this.
For example, if it does not match JSON schema or something.

**Reply by iksent** [reply]
It's much more comfortable for developers to get the result object in places, where they expected to get a single item.
For example, when retrieving the article by id or by slug on a single article page.
IMHO it's something strange to get an array and then extract the first item of it.

**Comment by rijkvanzanten** [comment]
This'll be implemented and is tracked in https://github.com/directus/directus/issues/4627
